src/rgw/rgw_common.cc

```diff
@@ -1167,13 +1167,9 @@ bool verify_user_permission(const DoutPrefixProvider* dpp,
     return true;
   }
 
-  if (op == rgw::IAM::s3CreateBucket || op == rgw::IAM::s3ListAllMyBuckets) {
```

Comment:
@yuvalif : can you help me understand why the check for CreateBucket/ListAllMyBuckets has been removed?

Reply:
@pritha-srivastava after my changes, we reach this line when we do bucket list, and return false.
why did we add this "if" statement i nthe first place?

Reply:
@yuvalif : iirc, I added that because if a user tries to create a bucket/ list all buckets and there wasn't any policy (user policy, or role policy or session policy) for the same, or the policy didn't have any action to allow these s3 ops, then the code would go to the last statement and return false and these ops would be denied, which shouldn't be. I added the check as part of this commit: https://github.com/ceph/ceph/pull/21379/commits/acee911064bf211f1e13a534ac7a226f693cae5b

Reply:
before my change, the code was doing the check only for bucket creation and listing all buckets.
so, for these operations nothing will change.
the change is that the chek will be done for other operation.

Reply:
But for other iam ops (like create oidc provider etc) the ops must be allowed only based on the evaluation of the policies and in the absence of these policies, the op must be denied.

Reply:
ok. i'll check what is happening in case of iam.
note that the point of the refactoring is that we don't have one generic "S3 service" handler to cover for: bucketless ops, IAM, STS and topics. instead, there should be a specific handler for each one (there already was, onlt it wasn't alsway used...) and we should avoid ops specific "if" statements there

Reply:
in that case, special handling can be added to s3CreateBucket and s3ListAllMyBuckets verify_permission() probably, since verify_user_permission is common to these s3 ops and all other iam, sts ops.

Reply:
ok. so in the "common" function I should just return "false" if there is no policy?
and add the policy-less check to the `verify_permission()` of `s3CreateBucket` and `s3ListAllMyBuckets` ?

Reply:
actually if verify_user_permission() returns false we wouldn't know whether its an actual 'deny' or simply because the policies don't exist or the policy doesn't have an action for these ops. So a better option would be to change the return type of the current verify_user_permission() to reflect the return type of policy evaluation, and then have a wrapper verify_user_permission with return type bool, that can invoke this new verify_user_permission. In this case we can move the extra 'if' check part to CreateBucket and ListAllMyBuckets verify_permission().

Reply:
@pritha-srivastava what about `RGWRestOIDCProvider::verify_permission()` is policy mandatory in this case or not?

Reply:
Policy is not mandatory in this case, a user with admin caps for oidc-provider can create an openidconnect provider in rgw.

Reply:
No, for OIDC ops we do not need to go to verify_user_permission_no_policy(). Is it not possible to invoke verify_user_permission_no_policy() in CreateBucket/ListAllMyBuckets(), after policy evaluation is complete and the result is an Effect::Pass?